common/txmgr/address_state.go

```diff
@@ -162,6 +163,60 @@ func (as *addressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) findTx
 	return as.idempotencyKeyToTx[key]
 }
 
+// addInProgressTxAttempt saves the in-progress transaction attempt.
+// The transaction attempt should have a valid ID assigned by the caller.
+func (as *addressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) addInProgressTxAttempt(
+	txAttempt txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
+) error {
+	as.Lock()
+	defer as.Unlock()
+
+	if txAttempt.ID == 0 {
+		return fmt.Errorf("save_in_progress_tx_attempt: TxAttempt ID must be set")
+	}
+
+	tx, ok := as.allTxs[txAttempt.TxID]
+	if !ok {
+		return ErrTxnNotFound
+	}
+
+	// add the new attempt to the transaction
+	tx.TxAttempts = append(tx.TxAttempts, txAttempt)
+	as.attemptHashToTxAttempt[txAttempt.Hash] = &txAttempt
+
+	return nil
+}
+
+// updateInProgressTxAttempt saves the in-progress transaction attempt.
+// The transaction attempt should have a valid ID assigned by the caller.
+func (as *addressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) updateInProgressTxAttempt(
+	txAttempt txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
+) error {
+	as.Lock()
+	defer as.Unlock()
+
+	if txAttempt.ID == 0 {
+		return fmt.Errorf("save_in_progress_tx_attempt: TxAttempt ID must be set")
+	}
+
+	tx, ok := as.allTxs[txAttempt.TxID]
+	if !ok {
+		return ErrTxnNotFound
+	}
+
+	// update the existing attempt if it exists
+	for i := 0; i < len(tx.TxAttempts); i++ {
+		if tx.TxAttempts[i].ID == txAttempt.ID {
+			tx.TxAttempts[i].State = txmgrtypes.TxAttemptInProgress
+			tx.TxAttempts[i].BroadcastBeforeBlockNum = txAttempt.BroadcastBeforeBlockNum
+			as.attemptHashToTxAttempt[txAttempt.Hash] = &tx.TxAttempts[i]
+			return nil
+		}
+	}
```

Comment:
If tx.TxAttemts was `[]*TxAttempt` instead we could make use of the hashing. This may help optimize other sections as well.

```suggestion
	if attempt, ok := as.attemptHashToTxAttempt[txAttempt.Hash]; ok {
		*attempt = txAttempt
		attempt.State = txmgrtypes.TxAttemptInProgress
		return nil
	}
```